src/main/java/com/harena/api/service/PatrimoineService.java

```diff
@@ -1,72 +1,106 @@
+import com.harena.api.file.BucketConf;
+import com.harena.api.utils.PageRequest;
 import com.harena.api.utils.StringNormalizer;
 import java.io.File;
+import java.io.IOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.util.HashMap;
+import java.util.ArrayList;
 import java.util.List;
-import java.util.Optional;
 import lombok.RequiredArgsConstructor;
-import lombok.SneakyThrows;
 import lombok.extern.java.Log;
 import org.springframework.stereotype.Service;
 import school.hei.patrimoine.modele.Patrimoine;
 import school.hei.patrimoine.serialisation.Serialiseur;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Request;
+import software.amazon.awssdk.services.s3.model.ListObjectsV2Response;
+import software.amazon.awssdk.services.s3.paginators.ListObjectsV2Iterable;
 
 @Log
 @Service
 @RequiredArgsConstructor
 public class PatrimoineService {
-  private final String PATRIMOINE_KEY = "patrimoines.db";
+  private final String PATRIMOINE_KEY_PREFIX = "patrimoines/";
   private final BucketComponent bucketComponent;
-  private final Serialiseur<HashMap<String, Patrimoine>> serializer = new Serialiseur<>();
+  private final BucketConf bucketConf;
+  private final Serialiseur<Patrimoine> serializer = new Serialiseur<>();
 
-  @SneakyThrows
-  public Optional<Patrimoine> getPatrimone(String patrimoineName) {
-    String id = StringNormalizer.apply(patrimoineName);
-    File patrimoineFile = bucketComponent.download(PATRIMOINE_KEY);
-    if (!patrimoineFile.exists()) {
-      return Optional.empty();
+  public Patrimoine getPatrimone(String patrimoineName) {
+    File patrimoineFile;
+    try {
+      patrimoineFile =
+          bucketComponent.download(PATRIMOINE_KEY_PREFIX + StringNormalizer.apply(patrimoineName));
+    } catch (Exception e) {
+      throw new NotFoundException(String.format("Patrimoine %s not found", patrimoineName));
     }
 
-    HashMap<String, Patrimoine> patrimoine =
-        serializer.deserialise(Files.readString(patrimoineFile.toPath()));
-    return Optional.ofNullable(patrimoine.get(id));
+    try {
+      return serializer.deserialise(Files.readString(patrimoineFile.toPath()));
+    } catch (IOException | ClassCastException e) {
+      throw new InternalServerErrorException(
+          String.format("Could not read values of patrimoine %s", patrimoineName));
+    }
   }
 
-  @SneakyThrows
   public List<Patrimoine> savePatrimoines(List<Patrimoine> toSavePatrimoines) {
-    File oldPatrimoinesFile = bucketComponent.download(PATRIMOINE_KEY);
-    if (!oldPatrimoinesFile.exists()) {
-      oldPatrimoinesFile.createNewFile();
-      HashMap<String, Patrimoine> emptyPatrimoine = new HashMap<>();
-      Files.writeString(oldPatrimoinesFile.toPath(), serializer.serialise(emptyPatrimoine));
+    try {
+      Path tempDir = Files.createTempDirectory("harena-api-toSavePatrimoines");
+      toSavePatrimoines.forEach((patrimoine) -> writeFile(tempDir, patrimoine));
```

Comment:
superfluous ()

Reply:
Deep implications for the courageous ones: (1) partial application of function (2) curryfication